com.unity.sg2/Editor/GraphDeltaRegistry/FunctionDefinitions/NodeDescriptorNodeBuilder.cs

```diff
@@ -220,18 +220,21 @@ public void BuildNode(NodeHandler node, Registry registry)
             foreach (var param in functionDescriptor.Parameters)
             {
                 var shaderType = ShaderTypeForParameter(container, param, fallbackType);
-                if (param.Usage == Usage.In ||
-                    param.Usage == Usage.Static)
+
+                if (param.Usage == Usage.Out)
                 {
-                    shaderFunctionBuilder.AddInput(shaderType, param.Name);
+                    shaderFunctionBuilder.AddOutput(shaderType, param.Name);
                 }
-                else if(param.Usage == Usage.Local)
+                else if (param.Usage == GraphType.Usage.In
+                    || param.Usage == GraphType.Usage.Static
+                    || param.Usage == GraphType.Usage.Local && !ParametricTypeUtils.IsParametric(param))
                 {
-                    shaderFunctionBuilder.AddVariableDeclarationStatement(shaderType, param.Name);
+                    shaderFunctionBuilder.AddInput(shaderType, param.Name);
                 }
-                else if (param.Usage == Usage.Out)
+                else if(param.Usage == Usage.Local)
                 {
-                    shaderFunctionBuilder.AddOutput(shaderType, param.Name);
+                    var init = ParametricTypeUtils.ManagedToParametricToHLSL(param.DefaultValue);
+                    shaderFunctionBuilder.AddVariableDeclarationStatement(shaderType, param.Name, init);
```

Comment:
I changed the condition order, so that `Out` goes first, then `In/Static` and then `Local`. The diff is swiss cheese.

The real change here is on 236 for local usage, there is a util that tries to take the generic C# object stored on DefaultValue and pull out an hlsl initialization string based on the assumed type.

Reply:
Thank you for leaving this comment. The Swiss cheese diff was confusing me.